File: android/src/main/java/com/yodo1/yodo1_mas_flutter_plugin/Yodo1MasFlutterPlugin.java
package com.yodo1.yodo1_mas_flutter_plugin;

import android.app.Activity;
import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import com.yodo1.mas.Yodo1Mas;
import com.yodo1.mas.appopenad.Yodo1MasAppOpenAdListener;
import com.yodo1.mas.error.Yodo1MasError;
import com.yodo1.mas.helper.model.Yodo1MasAdBuildConfig;
import com.yodo1.mas.interstitial.Yodo1MasInterstitialAd;
import com.yodo1.mas.interstitial.Yodo1MasInterstitialAdListener;
import com.yodo1.mas.reward.Yodo1MasRewardAd;
import com.yodo1.mas.reward.Yodo1MasRewardAdListener;
import com.yodo1.mas.appopenad.Yodo1MasAppOpenAd;

import org.json.JSONException;
import org.json.JSONObject;

/** Yodo1MasFlutterPlugin */
public class Yodo1MasFlutterPlugin implements FlutterPlugin, MethodCallHandler, ActivityAware {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;
  private Activity activity;
  private Context context;

  private static final String CHANNEL = "com.yodo1.mas/sdk";
  private static final String METHOD_NATIVE_INIT_SDK = "native_init_sdk";
  private static final String METHOD_NATIVE_IS_AD_LOADED = "native_is_ad_loaded";
  private static final String METHOD_NATIVE_LOAD_AD = "native_load_ad";
  private static final String METHOD_NATIVE_SHOW_AD = "native_show_ad";
  private static final String METHOD_FLUTTER_INIT_EVENT = "flutter_init_event";
  private static final String METHOD_FLUTTER_AD_EVENT = "flutter_ad_event";
  private static final String AD_TYPE_REWARDED_NAME = "Rewarded";
  private static final String AD_TYPE_INTERSTITIAL_NAME = "Interstitial";
  private static final String AD_TYPE_APP_OPEN_NAME = "AppOpen";

  // Ad Type Codes
  static final int AD_TYPE_REWARD = 1;
  static final int AD_TYPE_INTERSTITIAL = 2;
  static final int AD_TYPE_APP_OPEN = 4;

  // Ad Event Codes
  static final int AD_EVENT_OPENED = 1001;
  static final int AD_EVENT_CLOSED = 1002;
  static final int AD_EVENT_FAILED_TO_OPEN = 1003;
  static final int AD_EVENT_LOADED = 1004;
  static final int AD_EVENT_FAILED_TO_LOAD = 1005;
  static final int AD_EVENT_EARNED = 2001;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), CHANNEL);
    channel.setMethodCallHandler(this);
    context = flutterPluginBinding.getApplicationContext();
    initializeEventListeners();
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    Log.d("Yodo1MasFlutterPlugin", "onMethodCall: " + call.method);
    switch (call.method) {
      case METHOD_NATIVE_INIT_SDK:
        handleInitSdk(call, result);
        break;
      case METHOD_NATIVE_LOAD_AD:
        handleLoadAd(call);
        result.success(null);
        break;
      case METHOD_NATIVE_IS_AD_LOADED:
        boolean isAdLoaded = handleIsAdLoaded(call);
        result.success(isAdLoaded);
        break;
      case METHOD_NATIVE_SHOW_AD:
        handleShowAd(call);
        result.success(null);
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  private void handleInitSdk(MethodCall call, Result result) {
    String appKey = call.argument("app_key");
    boolean privacy = call.argument("privacy");
    boolean ccpa = call.argument("ccpa");
    boolean coppa = call.argument("coppa");
    boolean gdpr = call.argument("gdpr");

    if (appKey != null) {
      Yodo1MasAdBuildConfig config = new Yodo1MasAdBuildConfig.Builder().enableUserPrivacyDialog(privacy).build();
      Yodo1Mas.getInstance().setCCPA(ccpa);
      Yodo1Mas.getInstance().setCOPPA(coppa);
      Yodo1Mas.getInstance().setGDPR(gdpr);
      Yodo1Mas.getInstance().setAdBuildConfig(config);
      Yodo1Mas.getInstance().initMas(activity, appKey, new Yodo1Mas.InitListener() {
        @Override
        public void onMasInitSuccessful() {
          JSONObject initEvent = new JSONObject();
          try {
            initEvent.put("successful", true);
          } catch (JSONException e) {
            e.printStackTrace();
          }
          Log.d("Yodo1MasFlutterPlugin", "Flutter SDK Callback - Init Successful - " + initEvent.toString());
          channel.invokeMethod(METHOD_FLUTTER_INIT_EVENT, initEvent.toString());
          result.success(1);
        }

        @Override
        public void onMasInitFailed(@NonNull Yodo1MasError error) {
          JSONObject initEvent = new JSONObject();
          try {
            initEvent.put("successful", false);
            initEvent.put("error", error.getJsonObject());
          } catch (JSONException e) {
            e.printStackTrace();
          }
          Log.d("Yodo1MasFlutterPlugin", "Flutter SDK Callback - Init Failed - " + initEvent.toString());
          channel.invokeMethod(METHOD_FLUTTER_INIT_EVENT, initEvent.toString());
          result.error("INIT_FAILED", error.getMessage(), null);
        }
      });

    } else {
      result.error("INVALID_ARGUMENTS", "Invalid arguments for native_init_sdk", null);
    }
  }

  private void handleLoadAd(MethodCall call) {
    String type = call.argument("ad_type");
    if (type != null) {
      Log.d("Yodo1MasFlutterPlugin", "Flutter SDK Channel Call - Load Ad - " + type);
      switch (type) {
        case Yodo1MasFlutterPlugin.AD_TYPE_REWARDED_NAME:
          Log.d("Yodo1MasFlutterPlugin", "Flutter SDK Channel Call - Loading Rewarded Ad - " + type);
          activity.runOnUiThread(() -> {
            Yodo1MasRewardAd.getInstance().autoDelayIfLoadFail = true;
            Yodo1MasRewardAd.getInstance().loadAd(activity);
          });
          break;
        case Yodo1MasFlutterPlugin.AD_TYPE_INTERSTITIAL_NAME:
          Log.d("Yodo1MasFlutterPlugin", "Flutter SDK Channel Call - Loading Interstitial Ad - " + type);
          activity.runOnUiThread(() -> {
            Yodo1MasInterstitialAd.getInstance().autoDelayIfLoadFail = true;
            Yodo1MasInterstitialAd.getInstance().loadAd(activity);
          });
          break;
        case Yodo1MasFlutterPlugin.AD_TYPE_APP_OPEN_NAME:
          Log.d("Yodo1MasFlutterPlugin", "Flutter SDK Channel Call - Loading AppOpenAd Ad - " + type);
          activity.runOnUiThread(() -> {
            Yodo1MasAppOpenAd.getInstance().autoDelayIfLoadFail = true;
            Yodo1MasAppOpenAd.getInstance().loadAd(activity);
          });
          break;
        default:
          break;
      }
    }
  }

  private boolean handleIsAdLoaded(MethodCall call) {
    String type = call.argument("ad_type");
    Log.d("Yodo1MasFlutterPlugin", "handleIsAdLoaded call arguments: " + call.arguments);
    if (type != null) {
      switch (type) {
        case Yodo1MasFlutterPlugin.AD_TYPE_REWARDED_NAME:
          return Yodo1MasRewardAd.getInstance().isLoaded();
        case Yodo1MasFlutterPlugin.AD_TYPE_INTERSTITIAL_NAME:
          return Yodo1MasInterstitialAd.getInstance().isLoaded();
        case Yodo1MasFlutterPlugin.AD_TYPE_APP_OPEN_NAME:
          return Yodo1MasAppOpenAd.getInstance().isLoaded();
        default:
          return false;
      }
    }
    return false;
  }

  private void handleShowAd(MethodCall call) {
    String type = call.argument("ad_type");
    String placementId = call.argument("placement_id");
    Log.d("Yodo1MasFlutterPlugin", "handleIsAdLoaded call arguments: " + call.arguments);
    if (type != null) {
      switch (type) {
        case Yodo1MasFlutterPlugin.AD_TYPE_REWARDED_NAME:
          activity.runOnUiThread(() -> {
            if (placementId != null) {
              Yodo1MasRewardAd.getInstance().showAd(activity, placementId);
            } else {
              Yodo1MasRewardAd.getInstance().showAd(activity);
            }
          });
          break;
        case Yodo1MasFlutterPlugin.AD_TYPE_INTERSTITIAL_NAME:
          activity.runOnUiThread(() -> {
            if (placementId != null) {
              Yodo1MasInterstitialAd.getInstance().showAd(activity, placementId);
            } else {
              Yodo1MasInterstitialAd.getInstance().showAd(activity);
            }
          });
          break;
        case Yodo1MasFlutterPlugin.AD_TYPE_APP_OPEN_NAME:
          activity.runOnUiThread(() -> {
            if (placementId != null) {
              Yodo1MasAppOpenAd.getInstance().showAd(activity, placementId);
            } else {
              Yodo1MasAppOpenAd.getInstance().showAd(activity);
            }
          });
          break;
        default:
          break;
      }
    }
  }

  private void initializeEventListeners() {
    // Initialize Rewarded Ad listeners
    Yodo1MasRewardAd.getInstance().setAdListener(new Yodo1MasRewardAdListener() {
      @Override
      public void onRewardAdLoaded(Yodo1MasRewardAd ad) {
        // Code to be executed when an ad finishes loading.
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_REWARD);
          adEvent.put("code", AD_EVENT_LOADED);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }


      @Override
      public void onRewardAdFailedToLoad(Yodo1MasRewardAd ad, @NonNull Yodo1MasError error) {
        // Code to be executed when an ad request fails.
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_REWARD);
          adEvent.put("code", AD_EVENT_FAILED_TO_LOAD);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }


      @Override
      public void onRewardAdOpened(Yodo1MasRewardAd ad) {
        // Code to be executed when an ad opens an overlay that
        // covers the screen.
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_REWARD);
          adEvent.put("code", AD_EVENT_OPENED);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }


      @Override
      public void onRewardAdFailedToOpen(Yodo1MasRewardAd ad, @NonNull Yodo1MasError error) {
        // Code to be executed when an ad open fails.
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_REWARD);
          adEvent.put("code", AD_EVENT_FAILED_TO_OPEN);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }


      @Override
      public void onRewardAdClosed(Yodo1MasRewardAd ad) {
        // Code to be executed when the user is about to return
        // to the app after tapping on an ad.
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_REWARD);
          adEvent.put("code", AD_EVENT_CLOSED);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }

      @Override
      public void onRewardAdEarned(Yodo1MasRewardAd ad) {
        // Code to be executed when the user is about to return
        // to the app after tapping on an ad.
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_REWARD);
          adEvent.put("code", AD_EVENT_EARNED);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }
    });
    // Initialize Interstitial Ad listeners
    Yodo1MasInterstitialAd.getInstance().setAdListener(new Yodo1MasInterstitialAdListener() {
      @Override
      public void onInterstitialAdLoaded(Yodo1MasInterstitialAd ad) {
        JSONObject adEvent = new JSONObject();
        Log.d("Yodo1MasFlutterPlugin", "onInterstitialAdLoaded 1");
        try {
          adEvent.put("type", AD_TYPE_INTERSTITIAL);
          adEvent.put("code", AD_EVENT_LOADED);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        Log.d("Yodo1MasFlutterPlugin", "onInterstitialAdLoaded 1" + METHOD_FLUTTER_AD_EVENT + " - " + adEvent.toString());
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }

      @Override
      public void onInterstitialAdFailedToLoad(Yodo1MasInterstitialAd ad, @NonNull Yodo1MasError error) {
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_INTERSTITIAL);
          adEvent.put("code", AD_EVENT_FAILED_TO_LOAD);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }

      @Override
      public void onInterstitialAdOpened(Yodo1MasInterstitialAd ad) {
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_INTERSTITIAL);
          adEvent.put("code", AD_EVENT_OPENED);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }

      @Override
      public void onInterstitialAdFailedToOpen(Yodo1MasInterstitialAd ad, @NonNull Yodo1MasError error) {
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_INTERSTITIAL);
          adEvent.put("code", AD_EVENT_FAILED_TO_OPEN);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }

      @Override
      public void onInterstitialAdClosed(Yodo1MasInterstitialAd ad) {
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_INTERSTITIAL);
          adEvent.put("code", AD_EVENT_CLOSED);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }
    });
    // Initialize App Open Ads listeners
    Yodo1MasAppOpenAd.getInstance().setAdListener(new Yodo1MasAppOpenAdListener() {
      @Override
      public void onAppOpenAdLoaded(Yodo1MasAppOpenAd ad) {
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_APP_OPEN);
          adEvent.put("code", AD_EVENT_LOADED);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }

      @Override
      public void onAppOpenAdFailedToLoad(Yodo1MasAppOpenAd ad, @NonNull Yodo1MasError error) {
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_APP_OPEN);
          adEvent.put("code", AD_EVENT_FAILED_TO_LOAD);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }

      @Override
      public void onAppOpenAdOpened(Yodo1MasAppOpenAd ad) {
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_APP_OPEN);
          adEvent.put("code", AD_EVENT_OPENED);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }

      @Override
      public void onAppOpenAdFailedToOpen(Yodo1MasAppOpenAd ad, @NonNull Yodo1MasError error) {
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_APP_OPEN);
          adEvent.put("code", AD_EVENT_FAILED_TO_OPEN);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }

      @Override
      public void onAppOpenAdClosed(Yodo1MasAppOpenAd ad) {
        JSONObject adEvent = new JSONObject();
        try {
          adEvent.put("type", AD_TYPE_APP_OPEN);
          adEvent.put("code", AD_EVENT_CLOSED);
        } catch (JSONException e) {
          e.printStackTrace();
        }
        channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, adEvent.toString());
      }
    });
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }

  @Override
  public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
    activity = binding.getActivity();
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {

  }

  @Override
  public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {

  }

  @Override
  public void onDetachedFromActivity() {

  }
}

File: ios/Classes/Yodo1MasFlutterPlugin.swift
import Flutter
import UIKit
import Yodo1MasCore
import YYModel

public class Yodo1MasFlutterPlugin: NSObject, FlutterPlugin, Yodo1MasRewardAdDelegate, Yodo1MasInterstitialAdDelegate, Yodo1MasBannerAdDelegate {
    
    private var controller: UIViewController?
    private weak var channel: FlutterMethodChannel?
    
    private static let CHANNEL = "com.yodo1.mas/sdk"
    private let METHOD_NATIVE_INIT_SDK = "native_init_sdk"
    private let METHOD_NATIVE_IS_AD_LOADED = "native_is_ad_loaded"
    private let METHOD_NATIVE_LOAD_AD = "native_load_ad"
    private let METHOD_NATIVE_SHOW_AD = "native_show_ad"
    private let METHOD_FLUTTER_INIT_EVENT = "flutter_init_event"
    private let METHOD_FLUTTER_AD_EVENT = "flutter_ad_event"
    private let AD_TYPE_REWARDED_NAME = "Rewarded"
    private let AD_TYPE_INTERSTITIAL_NAME = "Interstitial"
    private let AD_TYPE_APP_OPEN_NAME = "AppOpen"
    private let AD_TYPE_BANNER_NAME = "Banner"
    private let AD_TYPE_NATIVE_NAME = "Native"

    // Ad Type Codes
    private let AD_TYPE_REWARDED = 1
    private let AD_TYPE_INTERSTITIAL = 2
    private let AD_TYPE_APP_OPEN = 3
    private let AD_TYPE_BANNER = 4
    private let AD_TYPE_NATIVE = 5

    // Event Type Codes
    private let AD_EVENT_OPENED = 1001
    private let AD_EVENT_CLOSED = 1002
    private let AD_EVENT_FAILED_TO_OPEN = 1003
    private let AD_EVENT_LOADED = 1004
    private let AD_EVENT_FAILED_TO_LOAD = 1005
    private let AD_EVENT_EARNED = 2001

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: Yodo1MasFlutterPlugin.CHANNEL, binaryMessenger: registrar.messenger())
        let instance = Yodo1MasFlutterPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case METHOD_NATIVE_INIT_SDK:
            handleInitSdk(call: call, result: result)
        case METHOD_NATIVE_LOAD_AD:
            handleLoadAd(call: call)
            result(nil)
        case METHOD_NATIVE_IS_AD_LOADED:
            let isAdLoaded = handleIsAdLoaded(call: call)
            result(isAdLoaded)
        case METHOD_NATIVE_SHOW_AD:
            handleShowAd(call: call)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleInitSdk(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let appKey = args["app_key"] as? String,
              let privacy = args["privacy"] as? Bool,
              let ccpa = args["ccpa"] as? Bool,
              let coppa = args["coppa"] as? Bool,
              let gdpr = args["gdpr"] as? Bool else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for native_init_sdk", details: nil))
            return
        }
        initSdk(appKey: appKey, privacy: privacy, ccpa: ccpa, coppa: coppa, gdpr: gdpr)
        result(1)
    }

    private func handleLoadAd(call: FlutterMethodCall) {
        guard let args = call.arguments as? [String: Any],
              let type = args["ad_type"] as? String else {
            return
        }
        switch type {
        case self.AD_TYPE_REWARDED_NAME:
            Yodo1MasRewardAd.sharedInstance().adDelegate = self
            Yodo1MasRewardAd.sharedInstance().autoDelayIfLoadFail = true
            Yodo1MasRewardAd.sharedInstance().load()
        case self.AD_TYPE_INTERSTITIAL_NAME:
            Yodo1MasInterstitialAd.sharedInstance().adDelegate = self
            Yodo1MasInterstitialAd.sharedInstance().autoDelayIfLoadFail = true
            Yodo1MasInterstitialAd.sharedInstance().load()
        case self.AD_TYPE_APP_OPEN_NAME:
            Yodo1MasInterstitialAd.sharedInstance().adDelegate = self
            Yodo1MasAppOpenAd.sharedInstance().autoDelayIfLoadFail = true
            Yodo1MasAppOpenAd.sharedInstance().load()
        default:
            break
        }
    }

    private func handleIsAdLoaded(call: FlutterMethodCall) -> Bool {
        guard let args = call.arguments as? [String: Any],
              let type = args["ad_type"] as? String else {
            return false
        }
        switch type {
        case self.AD_TYPE_REWARDED_NAME:
            return Yodo1MasRewardAd.sharedInstance().isLoaded()
        case self.AD_TYPE_INTERSTITIAL_NAME:
            return Yodo1MasInterstitialAd.sharedInstance().isLoaded()
        case self.AD_TYPE_APP_OPEN_NAME:
            return Yodo1MasAppOpenAd.sharedInstance().isLoaded()
        default:
            return false
        }
    }

    private func handleShowAd(call: FlutterMethodCall) {
        guard let args = call.arguments as? [String: Any],
              let type = args["ad_type"] as? String,
              let placementId = args["placement_id"] as? String? else {
            return
        }
        switch type {
        case self.AD_TYPE_REWARDED_NAME:
            if let placementId = placementId {
                Yodo1MasRewardAd.sharedInstance().show(withPlacement: placementId)
            } else {
                Yodo1MasRewardAd.sharedInstance().show()
            }
        case self.AD_TYPE_INTERSTITIAL_NAME:
            if let placementId = placementId {
                Yodo1MasInterstitialAd.sharedInstance().show(withPlacement: placementId)
            } else {
                Yodo1MasInterstitialAd.sharedInstance().show()
            }
        case self.AD_TYPE_APP_OPEN_NAME:
            if let placementId = placementId {
                Yodo1MasAppOpenAd.sharedInstance().show(withPlacement: placementId)
            } else {
                Yodo1MasAppOpenAd.sharedInstance().show()
            }
        default:
            break
        }
    }

    private func initSdk(appKey: String, privacy: Bool, ccpa: Bool, coppa: Bool, gdpr: Bool) {
        let config = Yodo1MasAdBuildConfig.instance()
        config.enableUserPrivacyDialog = privacy
        Yodo1Mas.sharedInstance().isCCPADoNotSell = ccpa
        Yodo1Mas.sharedInstance().isCOPPAAgeRestricted = coppa
        Yodo1Mas.sharedInstance().isGDPRUserConsent = gdpr
        Yodo1Mas.sharedInstance().setAdBuildConfig(config)
        
        Yodo1Mas.sharedInstance().initMas(withAppKey: appKey, successful: {
            self.channel?.invokeMethod(self.METHOD_FLUTTER_INIT_EVENT, arguments: ["successful": true])
        }, fail: { error in
            self.channel?.invokeMethod(self.METHOD_FLUTTER_INIT_EVENT, arguments: ["successful": false, "error": error.getJsonObject()])
        })
    }
    
    

}

extension Yodo1MasFlutterPlugin: Yodo1MasRewardDelegate {
    public func onRewardAdLoaded(_ ad: Yodo1MasRewardAd) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_REWARDED, "code": AD_EVENT_LOADED])
    }

    public func onRewardAdFailed(toLoad ad: Yodo1MasRewardAd, withError error: Yodo1MasError) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_REWARDED, "code": AD_EVENT_FAILED_TO_LOAD])
    }

    public func onRewardAdOpened(_ ad: Yodo1MasRewardAd) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_REWARDED, "code": AD_EVENT_OPENED])
    }

    public func onRewardAdFailed(toOpen ad: Yodo1MasRewardAd, withError error: Yodo1MasError) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_REWARDED, "code": AD_EVENT_FAILED_TO_OPEN])
    }

    public func onRewardAdClosed(_ ad: Yodo1MasRewardAd) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_REWARDED, "code": AD_EVENT_CLOSED])
    }

    public func onRewardAdEarned(_ ad: Yodo1MasRewardAd) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_REWARDED, "code": AD_EVENT_EARNED])
    }
}

extension Yodo1MasFlutterPlugin: Yodo1MasInterstitialDelegate {
    public func onInterstitialAdLoaded(_ ad: Yodo1MasInterstitialAd) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_INTERSTITIAL, "code": AD_EVENT_LOADED])
    }

    public func onInterstitialAdFailed(toLoad ad: Yodo1MasInterstitialAd, withError error: Yodo1MasError) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_INTERSTITIAL, "code": AD_EVENT_FAILED_TO_LOAD])
    }

    public func onInterstitialAdOpened(_ ad: Yodo1MasInterstitialAd) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_INTERSTITIAL, "code": AD_EVENT_OPENED])
    }

    public func onInterstitialAdFailed(toOpen ad: Yodo1MasInterstitialAd, withError error: Yodo1MasError) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_INTERSTITIAL, "code": AD_EVENT_FAILED_TO_LOAD])
    }

    public func onInterstitialAdClosed(_ ad: Yodo1MasInterstitialAd) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_INTERSTITIAL, "code": AD_EVENT_CLOSED])
    }
}

extension Yodo1MasFlutterPlugin: Yodo1MasAppOpenAdDelegate {
    public func onAppOpenAdLoaded(_ ad: Yodo1MasAppOpenAd) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_APP_OPEN, "code": AD_EVENT_LOADED])
    }

    public func onAppOpenAdFailed(toLoad ad: Yodo1MasAppOpenAd, withError error: Yodo1MasError) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_APP_OPEN, "code": AD_EVENT_FAILED_TO_LOAD])
    }

    public func onAppOpenAdOpened(_ ad: Yodo1MasAppOpenAd) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_APP_OPEN, "code": AD_EVENT_OPENED])
    }

    public func onAppOpenAdFailed(toOpen ad: Yodo1MasAppOpenAd, withError error: Yodo1MasError) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_APP_OPEN, "code": AD_EVENT_FAILED_TO_OPEN])
    }
    

    public func onAppOpenAdClosed(_ ad: Yodo1MasAppOpenAd) {
        channel?.invokeMethod(METHOD_FLUTTER_AD_EVENT, arguments: ["type": AD_TYPE_APP_OPEN, "code": AD_EVENT_CLOSED])
    }
}
